msu/systems/mod_settings/mod_settings_system.nut

```diff
@@ -0,0 +1,160 @@
+::MSU.Class.ModSettingsSystem <- class extends ::MSU.Class.System
+{
+	Panels = null;
+	Locked = null;
+
+	constructor()
+	{
+		base.constructor(::MSU.SystemID.ModSettings);
+		Locked = false;
+		this.Panels = ::MSU.Class.OrderedMap();
+	}
+
+	function add( _modPanel )
+	{
+		if (this.Locked)
+		{
+			::logError("Settings Manager is Locked, no more settings can be added");
+		}
+		else
+		{
+			if (!(_modPanel instanceof ::MSU.Class.SettingsPanel))
+			{
+				throw ::MSU.Exception.InvalidType;
+			}
+			this.Panels[_modPanel.getID()] <- _modPanel;
+		}
+	}
+
+	function get( _id )
+	{
+		return this.Panels[_id];
+	}
+
+	function getPanels()
+	{
+		return this.Panels;
+	}
+
+	function has( _id )
+	{
+		return this.Panels.contains(_id)
+	}
+
+	function finalize()
+	{
+		this.Locked = true;
+		this.Panels.sort(this.sortPanelsByName);
+	}
+
+	function updateSettings( _data, _informChange = true  )
+	{
+		/*
+		_data = {
+			modID = {
+				settingID = value
+			}
+		}
+		*/
+
+		foreach (modID, panel in _data)
+		{
+			foreach (settingID, value in panel)
+			{
+				local setting = this.Panels[modID].getSetting(settingID)
+				setting.set(value);
+			}
+		}
+	}
+
+	function setSettingFromPersistence( _modID, _settingID, _value )
+	{
+		if (!this.Panels.contains(_modID))
+		{
+			::MSU.Mod.Debug.printWarning(format("The mod %s has been removed", _modID), ::MSU.System.Debug.MSUMainDebugFlag);
+			return;
+		}
+		try
+		{
+			this.get(_modID).getSetting(_settingID).set(_value, true, false);
+		}
+		catch (error)
+		{
+			if (error == ::MSU.Exception.KeyNotFound)
+			{
+				::MSU.Mod.Debug.printWarning(format("Mod %s no longer has the setting %s", _modID, _settingID), ::MSU.System.Debug.MSUMainDebugFlag);
+				return;
+			}
+			throw error;
+		}
+	}
+
+	function callPanelsFunction(_function, ...)
+	{
+		vargv.insert(0, null);
+
+		foreach (panel in this.Panels)
+		{
+			vargv[0] = panel;
+			panel[_function].acall(vargv);
+		}
+	}
+
+	function importPersistentSettings()
+	{
+		::MSU.PersistentDataManager.loadSettingForEveryMod("ModSetting")
+	}
+
+	function flagSerialize()
+	{
+		this.callPanelsFunction("flagSerialize");
+	}
+
+	function resetFlags()
+	{
+		this.callPanelsFunction("resetFlags");
+	}
+
+	function flagDeserialize()
+	{
+		this.callPanelsFunction("flagDeserialize");
+	}
+
+	function getUIData( _flags = null )
+	{
+		local ret = [];
+		foreach (panel in this.Panels)
+		{
+			if (panel.verifyFlags(_flags))
+			{
+				ret.push(panel.getUIData(_flags));
+			}
+		}
+		ret.reverse() //No idea why conversion to JS reverses the array, but it does so I am pre-empting this.
+		return ret;
+	}
+
+	function getLogicalData()
+	{
+		local ret = {};
+		foreach (panel in this.getPanels())
+		{
+			ret[panel.getID()] <- panel.getLogicalData();
+		}
+		return ret;
+	}
+
+	function sortPanelsByName( _key1, _mod1, _key2, _mod2 )
```

Comment:
Actually not straightforward via sortAscending because the member is a function and we don't handle function members in sortAscending. Should we? That'll reduce the performance of that function because of a typeof check.

Reply:
Ok we can now handle functions in sortAscending but I still don't see how this can be done via that because Panels is an OrderedMap which expects a 4 argument function whereas sortAscending works with 2 arguments.